src/machine.rs

```diff
@@ -744,3 +760,29 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'mir, 'tcx> {
         res
     }
 }
+
+// This is potentially a performance hazard.
+// Factoring it into its own function lets us keep an eye on how much it shows up in a profile.
+fn set_current_span<'mir, 'tcx: 'mir>(machine: &Evaluator<'mir, 'tcx>) {
+    if let Some(sb) = machine.stacked_borrows.as_ref() {
+        if sb.borrow().current_span != DUMMY_SP {
```

Comment:
Is this related to the "lazy" aspect of span computation? Look like this 'cache' is invalidated on each step and eagerly filled in each machine hook that might call into Stacked Borrows. Please make sure that this overall system is described in a comment somewhere.